component/discovery/transformer/transformer.go

```diff
@@ -0,0 +1,212 @@
+
+// DecodeHCL implements gohcl.Decoder.
+// This method is only called on blocks, not objects.
+func (rc *RelabelConfig) DecodeHCL(body hcl.Body, ctx *hcl.EvalContext) error {
+	*rc = DefaultRelabelConfig
+
+	type relabelConfig RelabelConfig
+	err := gohcl.DecodeBody(body, ctx, (*relabelConfig)(rc))
+	if err != nil {
+		return err
+	}
+
+	if rc.Action == "" {
+		return fmt.Errorf("relabel action cannot be empty")
+	}
+	if rc.Modulus == 0 && rc.Action == HashMod {
+		return fmt.Errorf("relabel configuration for hashmod requires non-zero modulus")
+	}
+	if (rc.Action == Replace || rc.Action == HashMod || rc.Action == Lowercase || rc.Action == Uppercase) && rc.TargetLabel == "" {
+		return fmt.Errorf("relabel configuration for %s action requires 'target_label' value", rc.Action)
+	}
+	if (rc.Action == Replace || rc.Action == Lowercase || rc.Action == Uppercase) && !relabelTarget.MatchString(rc.TargetLabel) {
+		return fmt.Errorf("%q is invalid 'target_label' for %s action", rc.TargetLabel, rc.Action)
+	}
+	if (rc.Action == Lowercase || rc.Action == Uppercase) && rc.Replacement != DefaultRelabelConfig.Replacement {
+		return fmt.Errorf("'replacement' can not be set for %s action", rc.Action)
+	}
+	if rc.Action == LabelMap && !relabelTarget.MatchString(rc.Replacement) {
+		return fmt.Errorf("%q is invalid 'replacement' for %s action", rc.Replacement, rc.Action)
+	}
+	if rc.Action == HashMod && !model.LabelName(rc.TargetLabel).IsValid() {
+		return fmt.Errorf("%q is invalid 'target_label' for %s action", rc.TargetLabel, rc.Action)
+	}
+
+	if rc.Action == LabelDrop || rc.Action == LabelKeep {
+		if rc.SourceLabels != nil ||
+			rc.TargetLabel != DefaultRelabelConfig.TargetLabel ||
+			rc.Modulus != DefaultRelabelConfig.Modulus ||
+			rc.Separator != DefaultRelabelConfig.Separator ||
+			rc.Replacement != DefaultRelabelConfig.Replacement {
+
+			return fmt.Errorf("%s action requires only 'regex', and no other fields", rc.Action)
+		}
+	}
+
+	return nil
+}
+
+// Exports holds values which are exported by the discovery.transformer component.
+type Exports struct {
+	OutputTargets []Target `hcl:"output_targets,attr"`
+}
+
+// Component implements the discovery.transformer component.
+type Component struct {
+	opts component.Options
+
+	mut  sync.Mutex
+	args Arguments
+}
+
+var (
+	_ component.Component = (*Component)(nil)
+)
+
+// New creates a new discovery.transformer component.
+func New(o component.Options, args Arguments) (*Component, error) {
+	c := &Component{opts: o}
+
+	// Call to Update() to set the output once at the start
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	<-ctx.Done()
+	return nil
+}
+
+// Update implements component.Component.
+func (c *Component) Update(args component.Arguments) error {
+	newArgs := args.(Arguments)
+
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	c.args = newArgs
```

Comment:
c.args is never read, so it looks like you can remove both the mutex and that field :)

Reply:
You mean that we don't ever need it asynchronously so we can just use the component.Arguments passed in the Update() function, right?

Reply:
Nice catch, thanks!